src/hotstuff/voting.rs

```diff
@@ -0,0 +1,118 @@
+                select_leader(vote.view+1, validator_set_state.previous_validator_set()),
+            Phase::Decide => 
+                select_leader(vote.view+1, validator_set_state.committed_validator_set()),
+        }
+    }
+}
+
+/// Returns whether the replica with a given public key is allowed to vote for a nudge or proposal with
+/// a given justify.
+/// 
+/// Usually, a replica is allowed to vote if it belongs to the committed validator set. However,
+/// during the validator set update period replicas from the previous validator set are also
+/// allowed to vote for nudges and proposals for the validator-set-updating block, other than the
+/// "decide" phase nudge containing a commitQC. This is a fallback mechanism in case only selected
+/// replicas enter the validator set update period.
+/// 
+/// # Precondition
+/// justify satisfies [safe_qc](crate::state::safety::safe_qc) and 
+/// [is_correct](crate::hotstuff::types::QuorumCertificate::is_correct), and the block tree updates
+/// associated with this justify have already been applied.
+pub fn is_voter(validator: &VerifyingKey, validator_set_state: &ValidatorSetState, justify: &QuorumCertificate) -> bool {
+    if validator_set_state.update_completed() {
+        validator_set_state.committed_validator_set().contains(&validator)
+    } else {
+        match justify.phase {
+            Phase::Generic | Phase::Prepare | Phase::Precommit | Phase::Decide => 
+                validator_set_state.previous_validator_set().contains(&validator),
+            Phase::Commit => {
+                validator_set_state.committed_validator_set().contains(&validator)
+            }
+        }
+    }
+}
+
+/// Returns whether the replica with a given public key is an active validator. An active 
+/// validator can:
+/// 1. Propose/nudge and vote in the HotStuff protocol under certain circumstances described above,
+/// 2. Contribute [timeout votes](crate::pacemaker::messages::TimeoutVote) and 
+///    [advance view messages](crate::pacemaker::messages::AdvanceView).
+/// 
+/// In general, memebers of the committed validator set are always active, but during the validator
+/// set update period members of the previous validator set are active too.
+pub fn is_validator(verifying_key: &VerifyingKey, validator_set_state: &ValidatorSetState) -> bool {
+    validator_set_state.committed_validator_set().contains(verifying_key) ||
+    (!validator_set_state.update_completed() && 
+    validator_set_state.previous_validator_set().contains(verifying_key))
+}
+
+/// Returns the leader(s) of a given view. In general, a view has only one leader, but during the 
+/// validator set update period the leader of the resigning validator set can act as a leader too.
+pub fn leaders(view: ViewNumber, validator_set_state: &ValidatorSetState) -> (VerifyingKey, Option<VerifyingKey>) {
+    (select_leader(view, validator_set_state.committed_validator_set()),
```

Comment:
Definitely not something we can do for this release, but it'd be interesting to think about how we could maybe remove this coupling between HotStuff and Pacemaker.

Reply:
Agreed, I am also very dissatisfied with how coupled they are sometimes.